pdns/ixfrdist.cc

```diff
@@ -202,20 +202,22 @@ static bool sortSOA(uint32_t i, uint32_t j) {
 
 static void cleanUpDomain(const DNSName& domain, const uint16_t& keep, const string& workdir) {
   string dir = workdir + "/" + domain.toString();
-  DIR *dp;
-  dp = opendir(dir.c_str());
-  if (dp == nullptr) {
+  auto dirHandle = std::unique_ptr<DIR, decltype(&closedir)>(opendir(dir.c_str()), closedir);
+  if (!dirHandle) {
     return;
   }
   vector<uint32_t> zoneVersions;
-  struct dirent *d;
-  while ((d = readdir(dp)) != nullptr) {
-    if(!strcmp(d->d_name, ".") || !strcmp(d->d_name, "..")) {
+  struct dirent* entry = nullptr;
+  // NOLINTNEXTLINE(concurrency-mt-unsafe): readdir is thread-safe nowadays and readdir_r is deprecated
+  while ((entry = readdir(dirHandle.get())) != nullptr) {
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay): this is what dirent is
+    if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0) {
       continue;
     }
-    zoneVersions.push_back(std::stoi(d->d_name));
+    //  NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay): this is what dirent is
+    zoneVersions.push_back(std::stoi(entry->d_name));
   }
-  closedir(dp);
+  dirHandle.reset();
```

Comment:
Same here.

I feel this routine to open a directory and do something with the entries inside could be moved to the `pdns::` namespace in `misc.cc/hh`, and it could accept a lambda for the condition and/or a closure for taking some action on entries.

Reply:
Sounds like a good idea, I'll give it a try later today. If it leads to a big commit I might merge this PR first and open a new one.